### ISSUE TITLE
CORS problem when serving posix files on localhost

### ISSUE THREAD

**Issue by loichuder** [HDFGroup/hsds#28]
Hi there,

I have a local running instance of hsds serving POSIX files. 
Requests using the Python module _requests_ are succeeding. 

However, requests from a browser (using a React App running also in local in my case) fail due to CORS (_The Same Origin Policy disallows reading the remote resource_).

Any thoughts to resolve this ?

**Comment by loichuder** [comment]
Following up on this, I tried to set up CORS with the module [aiohttp-cors](https://github.com/aio-libs/aiohttp-cors), at least for the root of the domain (see https://github.com/loichuder/hsds/commit/da80064aca17b416b63fc66a5bb08c4d1cfbcefc).

The OPTIONS request passes smoothly but the GET request leads to an AssertionError thrown by aiohttp_cors in the service node (the same already reported in https://github.com/aio-libs/aiohttp-cors/issues/193)...

**Comment by loichuder** [comment]
After some digging, I saw that the CORS was in fact already manually handled in HSDS in the [`jsonResponse`](https://github.com/HDFGroup/hsds/blob/master/hsds/util/httpUtil.py#L284) function. Putting _aiohttp-cors_ on the top of that created a conflict which led to the error mentionned in my previous comment (https://github.com/HDFGroup/hsds/issues/28#issuecomment-598642264).

This being said, the issue still holds when using the HSDS `jsonReponse` CORS handling. I managed to solve it by removing the CORS in `jsonResponse` and using _aiohttp-cors_ instead (see https://github.com/loichuder/hsds/commit/39b13b93c1902d4bfb7159f0ef15417cbe2d192a).

@jreadey : what do you think ? Would it be better to use _aiohttp-cors_ then or am I in a special case?

**Comment by jreadey** [comment]
@loichuder - cool that you have the posix setup running!

Why was CORS failing (before you setup aiohttp-cors)?  I thought the headers the jsonRespnse sets should be adequate.  

I wasn't aware of the aiohttp-cors package.  Does it add something that would be useful?  I'm not oppose to using aiohttp-cors if that would be better (and you are willing to do a PR!).

In any case, it would be nice if the integration tests had a check for CORS.   I usually work with the Python client, so may miss if there's a CORS regression.

**Comment by jjaraalm** [comment]
@loichuder are you using chrome? [Chrome will not support CORS from localhost](https://bugs.chromium.org/p/chromium/issues/detail?id=67743). I am running the posix version on a dedicated server and have not had any issues with CORS yet.

**Comment by loichuder** [comment]
I don't know why CORS was failing although headers were apparently set adequately as you said. 
Perhaps this would need additional investigations (only tested with Firefox) but I guess there are two factors here:
- I am running everything on localhost.
- I am not using nginx which also implements its own CORS from what I saw from the config.

As for aiohttp-cors, it is only adding CORS support for aiohttp. To me, the advantage is that it is more modular and flexible than having the headers set in `jsonResponse` (also it means that OPTIONS requests need not to be manually implemented). Last but not least, it solved my problem :)

**Comment by jreadey** [comment]
@loichuder - would you mind submitting a PR with your aiohttp-cors setup?  Also it would be handy to have an integ test that verifies the cors response headers.  I'll do some adhoc testing on my side.

BTW, you might have noticed I took the nginx container out for docker-compose.yml and docker-compose.posix.yml.  It's been a bit problematic and I figure anyone whose looking at handling a large number of clients would want to use kubernetes anyway (k8s has it's own internal LB).

**Comment by jjaraalm** [comment]
Just to chime in, I reproduced a CORS issue on chrome, but it was due to trying to set specific headers. In my case it was `X-Hdf-domain` that was causing problems (since `Host` is not settable with `fetch()`). Using request options instead of headers solved the problem.

**Comment by jreadey** [comment]
Ah - so custom headers seem a bit problematic with CORS.  Fortunately looks like the only one we use is X-Hdf-domain and that can be replaced by using the domain param.

@loichuder - where you using X-Hdf-domain?

**Comment by loichuder** [comment]
> @loichuder - would you mind submitting a PR with your aiohttp-cors setup? Also it would be handy to have an integ test that verifies the cors response headers. I'll do some adhoc testing on my side.
> 
No problem. I can do that.
> BTW, you might have noticed I took the nginx container out for docker-compose.yml and docker-compose.posix.yml. It's been a bit problematic and I figure anyone whose looking at handling a large number of clients would want to use kubernetes anyway (k8s has it's own internal LB).

Yeah I noticed. In fact, in my early tests of POSIX, I removed it myself as it was easier for testing.

@jjaraalm Thanks for your input ! Well, I didn't have any fancy headers and I passed the domain as a parameter so no use of `X-Hdf-domain` as far as I know. I can check again but as I said, I feel that running everything in local might be the issue.

**Comment by jjaraalm** [comment]
Interesting, it `aiohttp-cors` fixes it, that's good, but it seems mysterious. I am also running every thing on localhost while working at home, both `hsds` and a development version of a web app. I tested both chrome and firefox without problems. This is what a request looks like on chrome:

```
GET /?domain=/shots/0 HTTP/1.1
Host: localhost
Connection: keep-alive
Cache-Control: max-age=0
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.132 Safari/537.36
Sec-Fetch-Dest: empty
authorization: Basic YWRtaW46YWRtaW4=
Accept: */*
Origin: http://localhost:3000
Sec-Fetch-Site: same-site
Sec-Fetch-Mode: cors
Referer: http://localhost:3000/
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9
```

and response headers
```

HTTP/1.1 200 OK
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET, POST, DELETE, PUT, OPTIONS
Access-Control-Allow-Headers: Content-Type, api_key, Authorization
Content-Type: application/json; charset=utf-8
Content-Length: 836
Date: Tue, 24 Mar 2020 14:19:55 GMT
Server: Python/3.7 aiohttp/3.6.2
```

**Comment by loichuder** [comment]
What do you use as authorization credentials ?

**Comment by jjaraalm** [comment]
For my test server, I'm using password auth with the defualt user/passwords (`admin`, `test_user1`, `test_user2`) etc...

My js calls look like:
```js
  async GET(request) {
    const url = `${this.endpoint}${request}?domain=${this.domain}`;
    const response =  await fetch(url, {
      method: 'GET',
      mode: 'cors',
      headers: this.Headers,
      cache: 'no-cache'
    });
    if (!response.ok) {
      throw new Error(`HSDS Error ${response.status} : ${response.statusText}`);
    }
    return await response.json();
  }
```
where headers are
```js
    this.Headers = new Headers();
    this.Headers.append("Authorization",
                        `Basic ${base64.encode(`${username}:${password}`)}`);
```
I've only tested in `fetch()`-capable browsers and haven't tested polyfills for IE yet.

**Comment by loichuder** [comment]
Thanks, I will give a try with your method. I was using [axios](https://github.com/axios/axios) so that might be misconfiguration from my side.

**Comment by loichuder** [comment]
Ok so I tried using the same js calls as you (with `fetch`) and it still doesn't work. 

I tried to access `/?domain=/home/test_user1/test/tall.h5` but my preflight request OPTIONS returns a `403 Forbidden` with the following error messages:
- on Firefox: 
`Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at http://localhost/?domain=/home/test_user1/test/tall.h5. (Reason: CORS header ‘Access-Control-Allow-Origin’ missing)`

- on Chrome: 
`Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource`.

It seems to me that my OPTIONS request is not handled by the part of HSDS that adds the CORS headers (`jsonReponse`).

**Comment by jjaraalm** [comment]
Hmm, I don't understand. Can you post the actual request/response?

Check your credentials and that your user actually has permission to access that domain. If I try to access with bad credentials, I get a 200 on the preflight, 401 on the response, but the same error message as you in the console. I don't get a 403 though... It looks like `Access-Control-Allow-Origin` is not being set on 401 and is inadvertently flagged as a CORS violation.

**Comment by loichuder** [comment]
Here are the request headers:
```
Host: localhost
User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:74.0) Gecko/20100101 Firefox/74.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate
Access-Control-Request-Method: GET
Access-Control-Request-Headers: authorization
Referer: http://localhost:3000/
Origin: http://localhost:3000
Connection: keep-alive
Pragma: no-cache
Cache-Control: no-cache
```
Tell me if you need something more as I would need to set up once again the version of HSDS without my CORS handling.

About credentials, there are working with no problem elsewhere (including with `aiohttp-cors`). However, there is indeed something weird: when looking at the logs of the service node, it appears that the user detected in preflight request is `default` (and not `test_user1`) who has indeed not the rights to access this domain. Could this be the cause ?

**Comment by jjaraalm** [comment]
Ah, yes that makes sense. I allow `default` read permissions on all domains. It looks like ACL checks are being done on OPTION requests which are always anonymous. `aiohttp-cors` does look to be the simplest solution.

**Comment by jreadey** [comment]
Hey,  I merged the aiohttp-cors PR (https://github.com/HDFGroup/hsds/pull/32)  to master but somehow the CORS headers are not getting added now (Python 3.8 incompat?).

I manually added the headers as a tmp hack to get the travis tests to pass: https://github.com/HDFGroup/hsds/commit/cdd452434a98fb12732a8e9fafc43917b7d3580a. 

@loichuder - Could you take a look and see what's up with aiohttp-cors?

**Comment by loichuder** [comment]
Yes, aiohttp-cors appears not to be Python3.8-ready yet. I will try to get a minimal example working in Python 3.8 to pinpoint the issues.

**Comment by loichuder** [comment]
Ok I have looked into this and here is my progress so far:
1. I managed to have a MWE for `aiohttp_cors` in Python 3.8 so that may not be the issue. 
1. The failing tests were on the requests `/about` and `/info`. Correct me if I am wrong, but I think the latter is implemented by the head_node. Which means that `aiohttp_cors` should also be set up for the head_node.
1. Even with `aiohttp_cors` set up for every node, I cannot get CORS headers in responses from GET requests made by the `requests` module of Python. This might be because [non-browser requests are handled differently by CORS}(https://stackoverflow.com/questions/43432743/will-cors-policy-prevent-resource-access-from-non-browser-requests) as CORS headers are present in the responses from browser requests. I could obtain the CORS headers by simulating the preflight request (OPTIONS) by adapting [this SO thread](https://stackoverflow.com/questions/43432743/will-cors-policy-prevent-resource-access-from-non-browser-requests?noredirect=1&lq=1)
1. As a aside, apparently, you cannot use `Access-Control-Allow-Origin: *` if `Allow-Credentials` is set to `true` (here is a link to [SO](https://stackoverflow.com/questions/19743396/cors-cannot-use-wildcard-in-access-control-allow-origin-when-credentials-flag-i?rq=1) but I could find others). I don't know yet if this comes into play.

To sum up, I would propose to:
- Implement CORS with `aiohttp_cors` for the head node
- Change the failing tests by firing an OPTIONS request instead of a GET.

**Comment by jreadey** [comment]
The /about and /info requests go to an SN node.  The head node would typically not be accessed externally, so don't think we don't need to deal with CORS on the head nodde right now.

Using Options rather than GET in the test sounds good.

From the SO, it seems like Access-Control-Allow-Origin needs to be a config rather than a hard-coded '*' if authentication is used.  I can do this.

What about binary responses? e.g. GET /datasets/<id>/value.   Do they get CORS?

**Comment by jreadey** [comment]
I've checked in code to have CORS_DOMAIN be a config option (defaults as '*'): https://github.com/HDFGroup/hsds/commit/3bfce39a201d87d1f8519cfda53684f380424411.

**Comment by loichuder** [comment]
> The /about and /info requests go to an SN node. The head node would typically not be accessed externally, so don't think we don't need to deal with CORS on the head nodde right now.
> 
All right. I misunderstood then.
> Using Options rather than GET in the test sounds good.
> 
Well, that is the only way I found to get CORS headers with requests from the Python `requests` package. I will implement this in another PR.
> What about binary responses? e.g. GET /datasets//value. Do they get CORS?
>
Again, not when doing them with `requests` but they get CORS when looking at requests made from the browser.

**Comment by loichuder** [comment]
Tests implemented in #38.
